File: intern-backend/src/main/java/com/internBackend/leave/service/LeaveService.java
package com.internBackend.leave.service;

import com.internBackend.leave.dto.LeaveDTO;
import com.internBackend.leave.entity.Leave;
import java.util.List;


public interface LeaveService {
    List<Leave> fetchAllLeaves();
    List<Leave> getAllLeavesByEmployeeId(Long employeeId);
    Long countAllLeaves();
    Long countAllApprovedLeaves();
    Long countAllApprovedLeavesByEmployeeId(Long employeeId);

    Leave fetchLeaveById(Long id);
    Leave createLeave(Leave leave);
    Leave updateLeave(Leave leave);
    String deleteLeave(Leave leave);
    Leave changeLeaveStatus(Long id);

    List<LeaveDTO> getPendingLeaves();
    Long countPendingLeaves();
    List<LeaveDTO> getTodayAndApprovedLeaves();
    Long countTodayAndApprovedLeaves();

    LeaveDTO convertToDTO(Leave leave);

    List<LeaveDTO> fetchAllLeaveDTOs();
    List<LeaveDTO> fetchAllLeaveDTOsByEmployeeId(Long employeeId);
    Long countAllLeavesByEmployeeId(Long employeeId);
}

File: intern-frontend/src/app/leave/service/leave.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from '@angular/common/http';
import {Observable} from 'rxjs';
import {Leave} from '../model/leave.model';
import {LeaveWithFullName} from '../dto/LeaveWithFullName.dto';
import {Position} from '../../employee-list/dto/Position.dto';

@Injectable({
  providedIn: 'root',
})
export class LeaveService {
  private leaveUrl: string = '/api/v1/leaves';
  private employeeUrl: string = '/api/v1/employees';

  constructor(private http: HttpClient) {
  }

  fetchAllLeaves(): Observable<Leave[]> {
    return this.http.get<Leave[]>(`${this.leaveUrl}`);
  }

  getAllLeavesByEmployeeId(employeeId: number): Observable<Leave[]> {
    return this.http.get<Leave[]>(`${this.leaveUrl}/employee/${employeeId}`);
  }

  createLeave(data: Leave): Observable<Leave> {
    return this.http.post<Leave>(`${this.leaveUrl}`, data);
  }


  updateLeave(data: Leave): Observable<Leave> {
    return this.http.put<Leave>(`${this.leaveUrl}/${data.id}`, data);
  }


  deleteLeave(id: number): Observable<{ message: string }> {
    return this.http.delete<{ message: string }>(`${this.leaveUrl}/${id}`);
  }

  changeLeaveStatus(id: number): Observable<Leave> {
    return this.http.get<Leave>(`${this.leaveUrl}/changeLeaveStatus/${id}`);
  }

  getPendingLeaves(): Observable<LeaveWithFullName[]> {
    return this.http.get<LeaveWithFullName[]>(`${this.leaveUrl}/pendingLeaves`);
  }

  countPendingLeaves(): Observable<number> {
    return this.http.get<number>(`${this.leaveUrl}/countPendingLeaves`);
  }

  getTodayAndApprovedLeaves(): Observable<LeaveWithFullName[]> {
    return this.http.get<LeaveWithFullName[]>(`${this.leaveUrl}/todayAndApprovedLeaves`);
  }

  countTodayAndApprovedLeaves(): Observable<number> {
    return this.http.get<number>(`${this.leaveUrl}/countTodayAndApprovedLeaves`);
  }

  getAllLeavesByEmployeeIdWithFullName(employeeId: number): Observable<LeaveWithFullName[]> {
    return this.http.get<LeaveWithFullName[]>(`${this.leaveUrl}/withFullName/${employeeId}`);
  }

  getAllLeavesWithFullName(): Observable<LeaveWithFullName[]> {
    return this.http.get<LeaveWithFullName[]>(`${this.leaveUrl}/withFullName`);
  }

  countAllLeavesByEmployeeId(employeeId: number): Observable<number> {
    return this.http.get<number>(`${this.leaveUrl}/count/${employeeId}`);
  }

  countAllLeaves(): Observable<number> {
    return this.http.get<number>(`${this.leaveUrl}/count`);
  }

  changeEmployeePositionByEmployeeId(employeeId: number, position: Position): Observable<boolean> {
    return this.http.put<boolean>(`${this.employeeUrl}/changeEmployeePosition/${employeeId}/${position}`, {});
  }

}
